### PR TITLE
[Feature] Run config for using edge features

### PR DESCRIPTION
**Pull request by zhjwy9343**
*Issue #, if available:*

*Description of changes:*
This PR is the last one to configure run main function in NP, EP, and LP. The major contents include:
1. Add `edge_feat_name` argument to Node, Edge, and LinkPrediction Dataloaders in NP, EP, and LP main() funcitons.
2. Add end-to-end test cases for NC, EC, ER, and LP.
3. Modify the `acm_data.py` example data generation code to demo using edge features.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: python/graphstorm/run/gsgnn_ep/gsgnn_ep.py [hunk 1]
```diff
@@ -151,6 +154,11 @@ def main(config_args):
                 grad_norm_type=config.grad_norm_type)
 
     if config.save_embed_path is not None:
+        assert config.edge_feat_name is None, 'GraphStorm edge prediction training command ' + \
+            'uses full-graph inference by default to generate node embeddings at the end of ' + \
+            'training. But the full-graph inference method does not support edge features ' + \
+            'in this version. Please set the \"save_embed_path\" to be empty or does not ' + \
```
Comment (classicsong) [inline python/graphstorm/run/gsgnn_ep/gsgnn_ep.py:160]:
```suggestion
            'in the current version. Please set the \"save_embed_path\" to none ' + \
```

File: python/graphstorm/run/gsgnn_ep/gsgnn_ep.py [hunk 2]
```diff
@@ -151,6 +154,11 @@ def main(config_args):
...
+            'set this argument if you want to use edge features in training command.'
```
Comment (classicsong) [inline python/graphstorm/run/gsgnn_ep/gsgnn_ep.py:161]:
```suggestion
            ' if you want to use edge features in training command.'
```

File: python/graphstorm/run/gsgnn_ep/gsgnn_lm_ep.py [hunk 3]
```diff
@@ -135,6 +138,11 @@ def main(config_args):
                 grad_norm_type=config.grad_norm_type)
 
     if config.save_embed_path is not None:
+        assert config.edge_feat_name is None, 'GraphStorm edge prediction training command ' + \
+            'uses full-graph inference by default to generate node embeddings at the end of ' + \
+            'training. But the full-graph inference method does not support edge features ' + \
+            'in this version. Please set the \"save_embed_path\" to be empty or does not ' + \
```
Comment (classicsong) [inline python/graphstorm/run/gsgnn_ep/gsgnn_lm_ep.py:144]: ditto

File: python/graphstorm/run/gsgnn_lp/gsgnn_lm_lp.py [hunk 4]
```diff
@@ -128,6 +133,11 @@ def main(config_args):
                 grad_norm_type=config.grad_norm_type)
 
     if config.save_embed_path is not None:
+        assert config.edge_feat_name is None, 'GraphStorm node prediction training command ' + \
+            'uses full-graph inference by default to generate node embeddings at the end of ' + \
+            'training. But the full-graph inference method does not support edge features ' + \
```
Comment (classicsong) [inline python/graphstorm/run/gsgnn_lp/gsgnn_lm_lp.py:138]: ditto

File: python/graphstorm/run/gsgnn_lp/gsgnn_lp.py [hunk 5]
```diff
@@ -146,6 +151,11 @@ def main(config_args):
                 grad_norm_type=config.grad_norm_type)
 
     if config.save_embed_path is not None:
+        assert config.edge_feat_name is None, 'GraphStorm node prediction training command ' + \
+            'uses full-graph inference by default to generate node embeddings at the end of ' + \
+            'training. But the full-graph inference method does not support edge features ' + \
+            'in this version. Please set the \"save_embed_path\" to be empty or does not ' + \
```
Comment (classicsong) [inline python/graphstorm/run/gsgnn_lp/gsgnn_lp.py:157]: ditto

File: python/graphstorm/run/gsgnn_np/gsgnn_np.py [hunk 6]
```diff
@@ -167,6 +170,11 @@ def main(config_args):
                 grad_norm_type=config.grad_norm_type)
 
     if config.save_embed_path is not None:
+        assert config.edge_feat_name is None, 'GraphStorm node prediction training command ' + \
+            'uses full-graph inference by default to generate node embeddings at the end of ' + \
+            'training. But the full-graph inference method does not support edge features ' + \
```
Comment (classicsong) [inline python/graphstorm/run/gsgnn_np/gsgnn_np.py:175]: ditto

File: tests/end2end-tests/graphstorm-ec/mgpu_test.sh [hunk 7]
```diff
@@ -342,3 +342,22 @@ python3 -m graphstorm.run.gs_edge_classification --workspace $GS_HOME/training_s
 error_and_exit $?
 
 rm -fr /tmp/*
+
+echo "**************dataset: MovieLens: EC, RGCN layer: 1, node feat: fixed HF BERT, BERT nodes: movie, edge feat: user,rating,movie:feat inference: mini-batch"
```
Comment (classicsong) [inline tests/end2end-tests/graphstorm-ec/mgpu_test.sh:346]:
```suggestion
echo "**************dataset: MovieLens: EC, RGCN layer: 1, node feat: fixed HF BERT, BERT nodes: movie, edge feat in message passing: user,rating,movie:feat inference: mini-batch"
```

File: tests/end2end-tests/graphstorm-ec/mgpu_test.sh [hunk 8]
```diff
@@ -342,3 +342,22 @@ python3 -m graphstorm.run.gs_edge_classification --workspace $GS_HOME/training_s
...
+python3 -m graphstorm.run.gs_edge_classification --workspace $GS_HOME/training_scripts/gsgnn_ep/ --num-trainers $NUM_TRAINERS --num-servers 1 --num-samplers 0 --part-config /data/movielen_100k_ef_nc_ec_train_val_1p_4t/movie-lens-100k.json --ip-config ip_list.txt --ssh-port 2222 --cf ml_ec.yaml --num-classes 5 --node-feat-name movie:title user:feat --edge-feat-name user,rating,movie:feat --batch-size 64 --save-model-path /data/gsgnn_ec_ml_ef/model/ --save-model-frequency 5 --eval-frequency 3  --num-epochs 1 --logging-file /tmp/train_log.txt  --backend nccl
+
```
Comment (classicsong) [inline tests/end2end-tests/graphstorm-ec/mgpu_test.sh:348]: Please add 
`error_and_exit $?`
to check the exit status.

File: tests/end2end-tests/graphstorm-ec/mgpu_test.sh [hunk 9]
```diff
@@ -342,3 +342,22 @@ python3 -m graphstorm.run.gs_edge_classification --workspace $GS_HOME/training_s
...
+python3 -m graphstorm.run.gs_edge_classification --inference --workspace $GS_HOME/inference_scripts/ep_infer/ --num-trainers $NUM_INFO_TRAINERS --num-servers 1 --num-samplers 0 --part-config /data/movielen_100k_ef_nc_ec_train_val_1p_4t/movie-lens-100k.json --ip-config ip_list.txt --ssh-port 2222 --cf ml_ec_infer.yaml --num-classes 5 --use-mini-batch-infer true --restore-model-path /data/gsgnn_ec_ml_ef/model/epoch-0/ --save-prediction-path /data/gsgnn_ec_ml_ef/prediction/ --logging-file /tmp/log.txt --preserve-input True --node-feat-name movie:title user:feat --edge-feat-name user,rating,movie:feat  --backend nccl
+
```
Comment (classicsong) [inline tests/end2end-tests/graphstorm-ec/mgpu_test.sh:350]: ditto

File: tests/end2end-tests/graphstorm-ec/mgpu_test.sh [hunk 10]
```diff
@@ -342,3 +342,22 @@ python3 -m graphstorm.run.gs_edge_classification --workspace $GS_HOME/training_s
...
+## Emb Gen
+python3 -m graphstorm.run.gs_gen_node_embedding --workspace $GS_HOME/training_scripts/gsgnn_ep/ --num-trainers $NUM_INFO_TRAINERS --num-servers 1 --num-samplers 0 --part-config /data/movielen_100k_ef_nc_ec_train_val_1p_4t/movie-lens-100k.json --ip-config ip_list.txt --ssh-port 2222 --cf ml_ec.yaml --num-classes 5 --node-feat-name movie:title user:feat --edge-feat-name user,rating,movie:feat --use-mini-batch-infer true --restore-model-path /data/gsgnn_ec_ml_ef/model/epoch-0/ --save-embed-path /data/gsgnn_ec_ml_ef/save-emb/ --logging-file /tmp/log.txt --logging-level debug  --backend nccl
+
```
Comment (classicsong) [inline tests/end2end-tests/graphstorm-ec/mgpu_test.sh:353]: ditto

File: tests/end2end-tests/graphstorm-er/test.sh [hunk 11]
```diff
@@ -7,7 +7,8 @@ GS_HOME=$(pwd)
 NUM_TRAINERS=1
 export PYTHONPATH=$GS_HOME/python/
 cd $GS_HOME/training_scripts/gsgnn_ep
-
+echo "127.0.0.1" > ip_list.txt
```
Comment (classicsong) [inline tests/end2end-tests/graphstorm-er/test.sh:10]:
```suggestion
```

File: tests/end2end-tests/graphstorm-er/test.sh [hunk 12]
```diff
@@ -7,7 +7,8 @@ GS_HOME=$(pwd)
...
+cd $GS_HOME/inference_scripts/ep_infer
```
Comment (classicsong) [inline tests/end2end-tests/graphstorm-er/test.sh:11]: Why change the workspace here?

File: tests/end2end-tests/graphstorm-er/test.sh [hunk 13]
```diff
@@ -71,6 +72,25 @@ python3 -m graphstorm.run.gs_edge_regression --workspace $GS_HOME/training_scrip
 
 error_and_exit $?
 
+echo "**************dataset: MovieLens: ER, RGCN layer: 1, node feat: fixed HF BERT, BERT nodes: movie, edge feat: user,rating,movie:feat inference: mini-batch"
+python3 -m graphstorm.run.gs_edge_regression --workspace $GS_HOME/training_scripts/gsgnn_ep/ --num-trainers $NUM_TRAINERS --num-servers 1 --num-samplers 0 --part-config /data/movielen_100k_ef_er_train_val_1p_4t/movie-lens-100k.json --ip-config ip_list.txt --ssh-port 2222 --cf ml_er.yaml --node-feat-name movie:title user:feat --edge-feat-name user,rating,movie:feat --batch-size 64 --save-model-path /data/gsgnn_er_ml_ef/model/ --save-model-frequency 5 --eval-frequency 3  --num-epochs 1 --logging-file /tmp/train_log.txt
+
```
Comment (classicsong) [inline tests/end2end-tests/graphstorm-er/test.sh:77]: Please add
error_and_exit $?
to check the exit status.

File: tests/end2end-tests/graphstorm-er/test.sh [hunk 14]
```diff
@@ -71,6 +72,25 @@ python3 -m graphstorm.run.gs_edge_regression --workspace $GS_HOME/training_scrip
...
+python3 -m graphstorm.run.gs_edge_regression --inference --workspace $GS_HOME/inference_scripts/ep_infer/ --num-trainers $NUM_TRAINERS --num-servers 1 --num-samplers 0 --part-config /data/movielen_100k_ef_er_train_val_1p_4t/movie-lens-100k.json --ip-config ip_list.txt --ssh-port 2222 --cf ml_er_infer.yaml --use-mini-batch-infer true --restore-model-path /data/gsgnn_er_ml_ef/model/epoch-0/ --save-prediction-path /data/gsgnn_er_ml_ef/prediction/ --logging-file /tmp/log.txt --preserve-input True --node-feat-name movie:title user:feat --edge-feat-name user,rating,movie:feat
+
```
Comment (classicsong) [inline tests/end2end-tests/graphstorm-er/test.sh:79]: ditto

File: tests/end2end-tests/graphstorm-er/test.sh [hunk 15]
```diff
@@ -71,6 +72,25 @@ python3 -m graphstorm.run.gs_edge_regression --workspace $GS_HOME/training_scrip
...
+## Emb Gen
+python3 -m graphstorm.run.gs_gen_node_embedding --workspace $GS_HOME/training_scripts/gsgnn_ep/ --num-trainers $NUM_TRAINERS --num-servers 1 --num-samplers 0 --part-config /data/movielen_100k_ef_er_train_val_1p_4t/movie-lens-100k.json --ip-config ip_list.txt --ssh-port 2222 --cf ml_er.yaml --node-feat-name movie:title user:feat --edge-feat-name user,rating,movie:feat --use-mini-batch-infer true --restore-model-path /data/gsgnn_er_ml_ef/model/epoch-0/ --save-embed-path /data/gsgnn_er_ml_ef/save-emb/ --logging-file /tmp/log.txt --logging-level debug
+
```
Comment (classicsong) [inline tests/end2end-tests/graphstorm-er/test.sh:82]: ditto

File: tests/end2end-tests/graphstorm-lp/mgpu_test.sh [hunk 16]
```diff
@@ -1052,3 +1052,21 @@ error_and_exit $?
 rm -fr /data/gsgnn_lp_ml_transe_l2/*
 
 rm -fr /tmp/*
+
+echo "**************dataset: MovieLens: LP, RGCN layer: 1, node feat: fixed HF BERT, BERT nodes: movie, edge feat: user,rating,movie:feat inference: mini-batch"
+python3 -m graphstorm.run.gs_link_prediction --workspace $GS_HOME/training_scripts/gsgnn_lp --num-trainers $NUM_TRAINERS --num-servers 1 --num-samplers 0 --part-config /data/movielen_100k_ef_lp_train_val_1p_4t/movie-lens-100k.json --ip-config ip_list.txt --ssh-port 2222 --cf ml_lp.yaml --num-epochs 1 --eval-frequency 30 --logging-file /tmp/train_log.txt --save-model-path /data/gsgnn_lp_ml_ef/model/  --save-model-frequency 50 --node-feat-name movie:title user:feat --edge-feat-name user,rating,movie:feat  --backend nccl
+
```
Comment (classicsong) [inline tests/end2end-tests/graphstorm-lp/mgpu_test.sh:1058]: Please add
error_and_exit $?
to check the exit status.

File: tests/end2end-tests/graphstorm-lp/mgpu_test.sh [hunk 17]
```diff
@@ -1052,3 +1052,21 @@ error_and_exit $?
...
+
+## Emb Gen
+python3 -m graphstorm.run.gs_gen_node_embedding --workspace $GS_HOME/training_scripts/gsgnn_lp --num-trainers $NUM_INFO_TRAINERS --num-servers 1 --num-samplers 0 --part-config /data/movielen_100k_ef_lp_train_val_1p_4t/movie-lens-100k.json --ip-config ip_list.txt --ssh-port 2222  --cf ml_lp.yaml --node-feat-name movie:title user:feat --edge-feat-name user,rating,movie:feat --restore-model-path /data/gsgnn_lp_ml_ef/model/epoch-0/ --save-embed-path /data/gsgnn_lp_ml_ef/save-emb/ --logging-file /tmp/log.txt --logging-level debug  --backend nccl
+
```
Comment (classicsong) [inline tests/end2end-tests/graphstorm-lp/mgpu_test.sh:1062]: Ditto

File: python/graphstorm/run/gsgnn_ep/gsgnn_lm_ep.py (hunk 3 above)
Comment (zhjwy9343) [reply python/graphstorm/run/gsgnn_ep/gsgnn_lm_ep.py:144]: changed.

File: python/graphstorm/run/gsgnn_lp/gsgnn_lm_lp.py (hunk 4 above)
Comment (zhjwy9343) [reply python/graphstorm/run/gsgnn_lp/gsgnn_lm_lp.py:138]: changed.

File: tests/end2end-tests/graphstorm-ec/mgpu_test.sh (hunk 8 above)
Comment (zhjwy9343) [reply tests/end2end-tests/graphstorm-ec/mgpu_test.sh:348]: added.

File: tests/end2end-tests/graphstorm-ec/mgpu_test.sh (hunk 9 above)
Comment (zhjwy9343) [reply tests/end2end-tests/graphstorm-ec/mgpu_test.sh:350]: added.

File: tests/end2end-tests/graphstorm-ec/mgpu_test.sh (hunk 10 above)
Comment (zhjwy9343) [reply tests/end2end-tests/graphstorm-ec/mgpu_test.sh:353]: added.

File: tests/end2end-tests/graphstorm-lp/mgpu_test.sh (hunk 16 above)
Comment (zhjwy9343) [reply tests/end2end-tests/graphstorm-lp/mgpu_test.sh:1058]: added

File: tests/end2end-tests/graphstorm-lp/mgpu_test.sh (hunk 17 above)
Comment (zhjwy9343) [reply tests/end2end-tests/graphstorm-lp/mgpu_test.sh:1062]: added.

File: tests/end2end-tests/graphstorm-er/test.sh (hunk 15 above)
Comment (zhjwy9343) [reply tests/end2end-tests/graphstorm-er/test.sh:81]: added.

File: tests/end2end-tests/graphstorm-er/test.sh (hunk 14 above)
Comment (zhjwy9343) [reply tests/end2end-tests/graphstorm-er/test.sh:78]: added.

File: tests/end2end-tests/graphstorm-er/test.sh (hunk 13 above)
Comment (zhjwy9343) [reply tests/end2end-tests/graphstorm-er/test.sh:76]: added.

File: tests/end2end-tests/graphstorm-er/test.sh (hunk 12 above)
Comment (zhjwy9343) [reply tests/end2end-tests/graphstorm-er/test.sh:10]: Other end2end tests use two workspaces. One for training and one for inference. And `ip_list.txt` file is created in both location.

But before this PR, there is only one workshop for ER tests. No previous case checks inference. As I need to do inference here, create a new `ip_list.txt` in the inference workspace.

**Review by classicsong — APPROVED** [review]

**Pull request merged**